### ISSUE TITLE
Singularity fails when launching a large number of instances with MPI

### ISSUE THREAD

**Issue by wangyinz** [apptainer/singularity#3928]
## Version of Singularity:

2.6.1, 3.2.1

### Expected behavior

Be able to start and run a mpi hello world code with a large number (>48) of tasks per node.

### Actual behavior

on 2.6.1 it may fail occasionally with either:
`ERROR : Failed to mount squashfs image in (read only): No such device`
or
`ERROR : No more available loop devices, try increasing 'max loop devices' in singularity.conf`
Note that these errors are more likely to occur when the program starts with no loop devices already in /dev. 

on 3.2.1 with shared loop devices enabled in the singularity.conf, it will fail with
`FATAL: container creation failed: mount /proc/self/fd/5->/var/singularity/mnt/session/rootfs error: can't mount image /proc/self/fd/5: failed to mount squashfs filesystem: no such device`
Note that all these errors are much less likely to appear when running with a small number of tasks (< 48) per node. I was actually not able to reproduce them at all with limited number of tests. 

### Steps to reproduce behavior

A very simple container with a hello world code should be enough to reproduce it:
`mpirun singularity exec container.simg hellow`

### Some thoughts

I think the error is from failure in attaching the image to available loop devices. By looking at the source code, I was only able to fix it with a lock placed at the beginning and end of the `_singularity_image_squashfs_mount` function in this [mount.c](https://github.com/sylabs/singularity/blob/2.6.1/src/lib/image/bind.c) file:

```
//Lock
singularity_message(DEBUG, "Trying to gain lock...\n");
int lock_fd;
remove("/tmp/singularity_loop_dev.lock");
while((lock_fd = open("/tmp/singularity_loop_dev.lock", O_CREAT | O_EXCL, 0600)) < 0) {
struct timespec ts_sleep;
ts_sleep.tv_sec = 0;
ts_sleep.tv_nsec = 100000000;
nanosleep(&ts_sleep, NULL);
}
singularity_message(DEBUG, "Lock Gained...\n");

...

//Unlock
if(close(lock_fd) < 0) {
singularity_message(ERROR, "Could not close lock file: %s\n", strerror(errno));
ABORT(255);
}
remove("/tmp/singularity_loop_dev.lock");
singularity_message(DEBUG, "Lock Released...\n");
```

This will sequencialize the loop device access, but it seems to be the only way to fix it. I also tried assigning different search range of loop devices according to the mpi_rank env variable, but that was not able to fix it. Also, putting the lock out side of this [loop](https://github.com/sylabs/singularity/blob/579c41508b4eb824c9f00e77a1fbf2b45fb7c637/src/lib/image/bind.c#L84) will only make the problem appear much less frequently, but I still occasionally get the "Failed to mount squashfs image" error.

Later on, I tested with 3.2.1 and found that this is actually fixed in 3.2.1 with a lock [here](https://github.com/sylabs/singularity/blob/b5e5078a95cee1657084b9e9aef105dd681d13aa/pkg/util/loop/loop_linux.go#L35). I have run ~50 tests, and never encountered a problem, but I am not 100% sure whether an error similar to that "Failed to mount squashfs image" may still exist, since the equivalent fix as I described above should in this [function](https://github.com/sylabs/singularity/blob/3bc1e0301155f3dff4343a3017f468ebc37102f8/internal/pkg/runtime/engines/singularity/container_linux.go#L566).  

While exploring all these, I found the second issue once enabled the shared loop devices option. It seems to be related the the first error message in 2.6.1. I was not able to look into this one yet, but I think it might worth reporting it here.

**Comment by wangyinz** [comment]
I just found another observation on the error of 3.2.1, which might be related. When the shared option is on, it seems Singularity will always create 256 loop devices (which is the default max) in `/dev` on both the successful and failed runs. This does not make sense as none of the tests are running with that many singularity instances.

**Comment by cclerget** [comment]
Hi @wangyinz , sorry the delayed response. Shared loop devices create missing loop devices because it requires to look at the 256 loop devices and the code create a loop block device when it can't open a `/dev/loopX`, it could be optimized but it was done intentionally to cover the case (very rare) where there is a deleted block device (eg: `/dev/loop100`) and `/dev/loop101` is actually in use.
Regarding the error `failed to mount squashfs filesystem: no such device` and according to `man 2 mount` it corresponds to `ENODEV: filesystemtype not configured in the kernel` which is quite surprising as error. The issue appears when image is stored on a shared filesystem or even on a local disk ?

**Comment by wangyinz** [comment]
Well, the image was on a shared file system, but I don't quite understand why the type of file system matters as I am producing these errors on a single node only but with many MPI tasks. The MPI tasks having conflict always share the same file system anyway.

**Comment by cclerget** [comment]
@wangyinz What's shared filesystem you use ? Even locally you call kernel VFS code of the network filesystem which may use a caching policy different from a local filesystem, so you may have issue that you won't have with a local filesystem. Are you able to trigger `failed to mount squashfs filesystem: no such device` when testing with an image on a local filesystem like `/tmp` ? Because I'm not able to reproduce the issue on a local filesystem.

**Comment by wangyinz** [comment]
That image was on a Lustre filesystem. I have not got the chance to test with local filesystem today. Will do it and report here. 

I thought it is only reading the image file, and do not know there is such as subtle difference that could become a problem.

**Comment by wangyinz** [comment]
I can confirm that the error still occurs for an image on a local file system. I have the image copied to `/tmp` before running the mpi code, and it still behaves the same.

Actually, based on what I have read from the source code and the debug info, I believe it is a conflict in accessing/creating the loop devices. The fix I provided in the original post also supports that suspect, as the loop device access is sequantialized in that fix.

**Comment by cclerget** [comment]
Was it a long running process executed by mpirun ? Can you try with 3.2.1 something basic like `mpirun singularity exec image sleep 10` if it always happens ?

In 3.X, access/creation of loop device can't conflict with other Singularity processes thanks to lock on `/dev` but there is a potential race (with `shared loop devices = yes`) where the first process associating the image with the loop device exit very quickly (which means the loop device is released at exit because of the autoclear flag set on the loop device) leading other processes to use an empty loop device. Maybe it could be that, a mpirun with a sleep of 10 seconds could confirm that., if you can give a try.

**Comment by wangyinz** [comment]
I have not done extensive tests with 3.X yet, but I can almost confirm that it has nothing to do with how long the process executes. The example I show is running an MPI hello world code, but when it fails, it never actually reached the MPI_Init call, so none of the processes should end.

**Comment by cclerget** [comment]
It will be hard to figure out what happens, that would be helpful to give it a try because the error `failed to mount squashfs filesystem: no such device` doesn't make any sense, that would mean that squashfs filesystem is not supported on your system but that's not the case as you shown. Now there is another error which could be returned is `failed to mount squashfs filesystem: no such device or address` and that would mean that you trigger the race I mentionned in comment https://github.com/sylabs/singularity/issues/3928#issuecomment-536688635 with 3.2.1 and `shared loop devices = yes`, that's why it would be greatly appreciated to test execution of `mpirun singularity exec image sleep 10` to confirm the issue disappears with a long running process in order to proceed with the right fix

**Comment by cclerget** [comment]
@wangyinz I submitted a fix #4549 for the race, if you have a chance to try it and let me know if that fixes the issue `failed to mount squashfs filesystem: no such device`. Thanks !

**Comment by cclerget** [comment]
Closing the issue as the fix was merged. Feel free to re-open it if the above fix doesn't work.
